File: src/ios/OSIABHiddenBrowserManager.swift
import Foundation
import WebKit

/// Manager for hidden InAppBrowser instances used for background authentication flows
class OSIABHiddenBrowserManager: NSObject {
    static let shared = OSIABHiddenBrowserManager()

    private var hiddenBrowsers: [String: HiddenBrowserInstance] = [:]

    private override init() {
        super.init()
    }

    class HiddenBrowserInstance: NSObject, WKNavigationDelegate {
        let webView: WKWebView
        let browserId: String
        var timer: Timer?
        let completionHandler: (OSIABEventType, Any?) -> Void
        private var firstLoadDone = false
        private let originalUrl: URL
        private var navigationCompletedTimer: DispatchWorkItem?
        private let navigationCompletedDelay: TimeInterval

        init(browserId: String, url: URL, options: OSIABWebViewOptions, customHeaders: [String: String]?, timeout: Int?, completionHandler: @escaping (OSIABEventType, Any?) -> Void) {
            self.browserId = browserId
            self.originalUrl = url
            self.navigationCompletedDelay = TimeInterval(options.navigationCompletedDelayMs) / 1000.0
            self.completionHandler = completionHandler

            // Create WKWebView configuration
            let configuration = WKWebViewConfiguration()
            configuration.allowsInlineMediaPlayback = options.allowInLineMediaPlayback
            configuration.mediaTypesRequiringUserActionForPlayback = options.mediaTypesRequiringUserActionForPlayback
            configuration.suppressesIncrementalRendering = options.surpressIncrementalRendering

            // Create a 1x1 hidden webview
            self.webView = WKWebView(frame: CGRect(x: 0, y: 0, width: 1, height: 1), configuration: configuration)

            super.init()

            // Configure WebView
            self.webView.navigationDelegate = self
            if let userAgent = options.customUserAgent {
                self.webView.customUserAgent = userAgent
            }

            // Clear cache if needed
            if options.clearCache {
                WKWebsiteDataStore.default().removeData(
                    ofTypes: WKWebsiteDataStore.allWebsiteDataTypes(),
                    modifiedSince: Date(timeIntervalSince1970: 0),
                    completionHandler: {}
                )
            } else if options.clearSessionCache {
                WKWebsiteDataStore.default().removeData(
                    ofTypes: Set([WKWebsiteDataTypeCookies, WKWebsiteDataTypeSessionStorage]),
                    modifiedSince: Date(timeIntervalSince1970: 0),
                    completionHandler: {}
                )
            }

            // Load URL with custom headers if provided
            var request = URLRequest(url: url)
            if let headers = customHeaders {
                for (key, value) in headers {
                    request.setValue(value, forHTTPHeaderField: key)
                }
            }
            self.webView.load(request)

            // Set up timeout if specified
            if let timeout = timeout, timeout > 0 {
                self.timer = Timer.scheduledTimer(withTimeInterval: TimeInterval(timeout), repeats: false) { [weak self] _ in
                    self?.completionHandler(.pageClosed, ["reason": "timeout"])
                }
            }
        }

        // WKNavigationDelegate methods
        func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
            // Cancel any pending navigation completed events since a new navigation is starting
            navigationCompletedTimer?.cancel()
            navigationCompletedTimer = nil
            decisionHandler(.allow)
        }

        /// Helper function to extract the domain (host) from a URL
        private func extractDomain(from url: URL?) -> String? {
            return url?.host
        }

        /// Helper function to check if a URL matches the original domain
        private func matchesOriginalDomain(_ url: URL?) -> Bool {
            guard let originalHost = extractDomain(from: originalUrl),
                  let currentHost = extractDomain(from: url) else {
                return false
            }
            return originalHost == currentHost
        }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            // Only fire pageLoadCompleted when we return to the original domain after potential SSO redirects
            if !firstLoadDone && matchesOriginalDomain(webView.url) {
                firstLoadDone = true
                completionHandler(.pageLoadCompleted, nil)
            } else {
                // Debounce the navigation completed event to handle redirect chains
                // Cancel any pending event first
                navigationCompletedTimer?.cancel()

                // Schedule new event to fire after delay (configurable)
                let workItem = DispatchWorkItem { [weak self] in
                    guard let self = self else { return }
                    self.completionHandler(.pageNavigationCompleted, self.webView.url?.absoluteString)
                    self.navigationCompletedTimer = nil
                }
                navigationCompletedTimer = workItem
                DispatchQueue.main.asyncAfter(deadline: .now() + navigationCompletedDelay, execute: workItem)
            }
        }

        func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
            completionHandler(.pageClosed, ["error": error.localizedDescription])
        }

        func cleanup() {
            timer?.invalidate()
            timer = nil
            navigationCompletedTimer?.cancel()
            navigationCompletedTimer = nil
            webView.stopLoading()
            webView.navigationDelegate = nil
        }

        deinit {
            cleanup()
        }
    }

    /// Creates and stores a hidden browser instance
    func create(browserId: String, url: URL, options: OSIABWebViewOptions, customHeaders: [String: String]?, timeout: Int?, completionHandler: @escaping (OSIABEventType, Any?) -> Void) {
        let instance = HiddenBrowserInstance(
            browserId: browserId,
            url: url,
            options: options,
            customHeaders: customHeaders,
            timeout: timeout,
            completionHandler: completionHandler
        )

        hiddenBrowsers[browserId] = instance
    }

    /// Removes a hidden browser instance
    func remove(browserId: String) {
        guard let instance = hiddenBrowsers[browserId] else { return }
        instance.cleanup()
        hiddenBrowsers.removeValue(forKey: browserId)
    }

    /// Gets a hidden browser instance
    func get(browserId: String) -> HiddenBrowserInstance? {
        return hiddenBrowsers[browserId]
    }

    /// Removes all hidden browsers
    func removeAll() {
        for (browserId, _) in hiddenBrowsers {
            remove(browserId: browserId)
        }
    }
}

File: src/android/OSIABHiddenBrowserManager.kt
package com.outsystems.plugins.inappbrowser.osinappbrowser

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import com.outsystems.plugins.inappbrowser.osinappbrowserlib.models.OSIABWebViewOptions
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Manager for hidden InAppBrowser instances used for background authentication flows
 */
object OSIABHiddenBrowserManager {
    private val hiddenBrowsers = ConcurrentHashMap<String, HiddenBrowserInstance>()

    /**
     * Hidden browser instance that runs in the background
     */
    class HiddenBrowserInstance(
        val browserId: String,
        context: Context,
        url: String,
        options: OSIABWebViewOptions,
        customHeaders: Map<String, String>?,
        timeout: Int?,
        val completionHandler: (OSIABEventType, Any?) -> Unit
    ) {
        private val webView: WebView
        private val handler = Handler(Looper.getMainLooper())
        private var timeoutRunnable: Runnable? = null
        private var navigationCompletedRunnable: Runnable? = null
        private var firstLoadDone = false
        private val originalUrl = url
        private val navigationCompletedDelayMs: Long = options.navigationCompletedDelayMs.toLong()

        /**
         * Helper function to extract the domain from a URL
         */
        private fun extractDomain(url: String?): String? {
            if (url == null) return null
            return try {
                val uri = Uri.parse(url)
                uri.host
            } catch (e: Exception) {
                null
            }
        }

        /**
         * Helper function to check if a URL matches the original domain
         */
        private fun matchesOriginalDomain(url: String?): Boolean {
            val originalDomain = extractDomain(originalUrl)
            val currentDomain = extractDomain(url)
            return originalDomain != null && currentDomain != null && originalDomain == currentDomain
        }

        init {
            webView = WebView(context).apply {
                // Configure WebView with options
                settings.apply {
                    javaScriptEnabled = true
                    domStorageEnabled = true
                    databaseEnabled = true
                    mediaPlaybackRequiresUserGesture = options.mediaPlaybackRequiresUserAction
                    allowFileAccess = true
                    allowContentAccess = true

                    // Set custom user agent if provided
                    options.customUserAgent?.let { userAgentString = it }
                }

                // Clear cache if needed
                if (options.clearCache) {
                    clearCache(true)
                    clearFormData()
                    clearHistory()
                }
                if (options.clearSessionCache) {
                    clearCache(false)
                }

                // Set WebViewClient to handle page events
                webViewClient = object : WebViewClient() {
                    override fun onPageStarted(view: WebView?, url: String?, favicon: android.graphics.Bitmap?) {
                        super.onPageStarted(view, url, favicon)
                        // Cancel any pending navigation completed events since a new navigation has started
                        navigationCompletedRunnable?.let { handler.removeCallbacks(it) }
                        navigationCompletedRunnable = null
                    }

                    override fun onPageFinished(view: WebView?, url: String?) {
                        // Only fire BROWSER_PAGE_LOADED when we return to the original domain after potential SSO redirects
                        if (!firstLoadDone && matchesOriginalDomain(url)) {
                            firstLoadDone = true
                            completionHandler(OSIABEventType.BROWSER_PAGE_LOADED, null)
                        } else {
                            // Debounce the navigation completed event to handle redirect chains
                            // Cancel any pending event first
                            navigationCompletedRunnable?.let { handler.removeCallbacks(it) }

                            // Schedule new event to fire after delay (configurable)
                            navigationCompletedRunnable = Runnable {
                                completionHandler(OSIABEventType.BROWSER_PAGE_NAVIGATION_COMPLETED, url)
                                navigationCompletedRunnable = null
                            }
                            handler.postDelayed(navigationCompletedRunnable!!, navigationCompletedDelayMs)
                        }
                    }

                    override fun onReceivedError(
                        view: WebView?,
                        request: WebResourceRequest?,
                        error: WebResourceError?
                    ) {
                        super.onReceivedError(view, request, error)
                        val errorData = mapOf("error" to (error?.description?.toString() ?: "Unknown error"))
                        completionHandler(OSIABEventType.BROWSER_FINISHED, errorData)
                    }
                }

                // Load URL with custom headers if provided
                if (customHeaders != null && customHeaders.isNotEmpty()) {
                    loadUrl(url, customHeaders)
                } else {
                    loadUrl(url)
                }
            }

            // Set up timeout if specified
            timeout?.let { timeoutSeconds ->
                if (timeoutSeconds > 0) {
                    timeoutRunnable = Runnable {
                        val timeoutData = mapOf("reason" to "timeout")
                        completionHandler(OSIABEventType.BROWSER_FINISHED, timeoutData)
                    }
                    handler.postDelayed(timeoutRunnable!!, (timeoutSeconds * 1000).toLong())
                }
            }
        }

        fun cleanup() {
            timeoutRunnable?.let { handler.removeCallbacks(it) }
            navigationCompletedRunnable?.let { handler.removeCallbacks(it) }
            webView.stopLoading()
            webView.destroy()
        }
    }

    /**
     * Creates and stores a hidden browser instance
     */
    fun create(
        browserId: String,
        context: Context,
        url: String,
        options: OSIABWebViewOptions,
        customHeaders: Map<String, String>?,
        timeout: Int?,
        completionHandler: (OSIABEventType, Any?) -> Unit
    ) {
        val instance = HiddenBrowserInstance(
            browserId,
            context,
            url,
            options,
            customHeaders,
            timeout,
            completionHandler
        )
        hiddenBrowsers[browserId] = instance
    }

    /**
     * Removes a hidden browser instance
     */
    fun remove(browserId: String) {
        hiddenBrowsers[browserId]?.let { instance ->
            instance.cleanup()
            hiddenBrowsers.remove(browserId)
        }
    }

    /**
     * Gets a hidden browser instance
     */
    fun get(browserId: String): HiddenBrowserInstance? {
        return hiddenBrowsers[browserId]
    }

    /**
     * Removes all hidden browsers
     */
    fun removeAll() {
        hiddenBrowsers.keys.forEach { browserId ->
            remove(browserId)
        }
    }
}
